public/app/plugins/datasource/prometheus/QueryCache.ts

```diff
@@ -0,0 +1,226 @@
+  targSigs: Map<TargetIdent, TargetSig>;
+  shouldCache: boolean,
+}
+
+export const getFieldIdent = (field: Field) => `${field.type}|${field.name}|${JSON.stringify(field.labels ?? '')}`;
+
+export class QueryCache {
+  cache = new Map<TargetIdent, TargetCache>();
+
+  // can be used to change full range request to partial, split into multiple requests
+  requestInfo(
+    request: DataQueryRequest<PromQuery>,
+    interpolateString: StringInterpolator
+  ): RequestInfo {
+    // TODO: align from/to to interval to increase probability of hitting backend cache
+
+    const newFrom = request.range.from.valueOf();
+    const newTo = request.range.to.valueOf();
+
+    // only cache 'now'-relative queries (that can benefit from a backfill cache)
+    const shouldCache = request.rangeRaw?.to?.toString() === 'now';
+
+    // all targets are queried together, so we check for any that causes group cache invalidation & full re-query
+    let doPartialQuery = shouldCache;
+    let prevTo: TimestampMs;
+
+    // pre-compute reqTargSigs
+    const reqTargSigs = new Map<TargetIdent, TargetSig>();
+    request.targets.forEach((targ) => {
+      let targIdent = `${request.dashboardUID}|${request.panelId}|${targ.refId}`;
+      let targExpr = interpolateString(targ.expr);
+      let targSig = `${targExpr}|${request.intervalMs}|${JSON.stringify(request.rangeRaw ?? '')}`; // ${request.maxDataPoints} ?
+
+      reqTargSigs.set(targIdent, targSig);
+    });
+
+    // figure out if new query range or new target props trigger full cache invalidation & re-query
+    for (const [targIdent, targSig] of reqTargSigs) {
+      let cached = this.cache.get(targIdent);
+      let cachedSig = cached?.sig;
+
+      if (cachedSig !== targSig) {
+        doPartialQuery = false;
+      } else {
+        // only do partial queries when new request range follows prior request range (possibly with overlap)
+        // e.g. now-6h with refresh <= 6h
+        prevTo = cached?.prevTo ?? Infinity;
+        doPartialQuery = newTo > prevTo && newFrom <= prevTo;
+      }
+
+      if (!doPartialQuery) {
+        break;
+      }
+    }
+
+    // console.log(`${doPartialQuery ? 'partial' : 'full'} query`);
+
+    if (doPartialQuery) {
+      // 10m requery overlap
+      const requeryLastMs = 10 * 60 * 1000;
+
+      // clamp to make sure we don't requery previous 10m when newFrom is ahead of it (e.g. 5min range, 30s refresh)
+      let newFromPartial = Math.max(prevTo! - requeryLastMs, newFrom);
+
+      // console.log(`query previous ${(newTo - newFromPartial) / 1000 / 60} mins`);
+
+      // align the query to the interval to increase the chance of hitting the backend cache
+      // ASK: how to test if this actually has desired effect?
+      // ASK: does the backend already do this anyways? do we need to do it here?
+      // newFromPartial = incrRoundDn(newFromPartial, request.intervalMs);
+      // newToPartial = incrRoundDn(newTo, request.intervalMs);
```

Comment:
I'll try to help with these questions. I need some more investigation.

Reply:
thanks!

it's not a huge deal to do it here. but would be good to know :)

Reply:
@itsmylife had you had a chance to look into these yet? These same questions are relevant to the cacheing PR we're working on as well.